cpp/src/arrow/dataset/partition.cc

```diff
@@ -646,15 +657,26 @@ class HivePartitioningFactory : public PartitioningFactory {
       }
     }
 
-    return impl.Finish(&dictionaries_);
+    auto schema_result = impl.Finish(&dictionaries_);
+    field_names_ = impl.FieldNames();
+    return schema_result;
   }
 
   Result<std::shared_ptr<Partitioning>> Finish(
       const std::shared_ptr<Schema>& schema) const override {
-    return std::shared_ptr<Partitioning>(new HivePartitioning(schema, dictionaries_));
+    for (FieldRef ref : field_names_) {
```

Comment:
I should probably guard here against the case that `field_names_` was not yet updated (if `Finish` is called without `Inspect` being called), with empty vector?

Reply:
Absolutely, the first line of this method should just call

```
auto field_names = FieldNames();
```

and replace occurrences of the private member.

Reply:
There is no `FieldNames()` method on the PartitioningFactory (only the `impl` has one, but that is not accessible here; that's the reason I added the `field_names_` private member to store those)